openedx/core/djangoapps/programs/tasks.py

```diff
@@ -0,0 +1,55 @@
+"""
+This file contains celery tasks for programs-related functionality.
+"""
+
+from celery import task
+from celery.utils.log import get_task_logger  # pylint: disable=no-name-in-module, import-error
+
+from certificates.api import get_certificates_for_user
+
+LOGGER = get_task_logger(__name__)
+
+
+@task
+def award_program_certificates(username):
+    """
+    This task is designed to be called whenever a user's completion status
+    changes with respect to one or more courses (primarily, when a course
+    certificate is awarded).
+
+    It will consult with a variety of APIs to determine whether or not the
+    specified user should be awarded a certificate in one or more programs, and
+    use the credentials service to create said certificates if so.
+
+    This task may also be invoked independently of any course completion status
+    change - for example, to backpopulate missing program credentials for a
+    user.
+
+    TODO: this is shelled out and incomplete for now.
+    """
+
+    # fetch the set of all course runs for which the user has earned a certificate
+    LOGGER.debug('fetching all completed courses for user %s', username)
+    user_certs = get_certificates_for_user(username)
+    course_certs = [
+        {'course_id': uc['course_id'], 'mode': uc['mode']}
```

Comment:
@rlucioni NOTE - this payload differs from what we discussed earlier - we need to pass the modes along with course ids, since that matters for program completion eligibility.